extensions/native/circuit/src/fri/mod.rs

```diff
@@ -204,17 +227,26 @@ impl FriReducedOpeningAir {
         let start_timestamp = next.general.timestamp;
         let multiplicity = local.prefix.general.is_workload_row;
         // a_ptr/b_ptr/length/result
-        let ptr_reads = AB::F::from_canonical_usize(4);
+        let ptr_reads = AB::F::from_canonical_usize(INSTRUCTION_READS);
         let native_as = AB::Expr::from_canonical_u32(AS::Native as u32);
-        // read a
+        // read a when write_a is 0
         self.memory_bridge
             .read(
+                MemoryAddress::new(native_as.clone(), next.data.a_ptr),
+                [local.prefix.a_or_is_first],
+                start_timestamp + ptr_reads,
+                local.a_aux.as_ref(),
+            )
+            .eval(builder, (AB::Expr::ONE - local_data.write_a) * multiplicity);
+        // write a when write_a is 1
+        self.memory_bridge
+            .write(
                 MemoryAddress::new(native_as.clone(), next.data.a_ptr),
                 [local.prefix.a_or_is_first],
                 start_timestamp + ptr_reads,
                 &local.a_aux,
             )
-            .eval(builder, multiplicity);
+            .eval(builder, local_data.write_a * multiplicity);
```

Comment:
assert `write_a` is bool? @zlangley

Reply:
`local_data.write_a = 1 - ood_point_idx`
the latter is currently **not** a boolean. we can fix it by either changing eDSL code or the constraints. probably eDSL is easier

the constraint `assert_bool` should be added

Reply:
@nyunyunyunyu I have defensively added https://github.com/openvm-org/openvm/pull/1347/commits/a008ab3ef109850207e410db7db90a1d771b9299 
Let me know if you agree / it is necessary